library/core/src/ptr/const_ptr.rs

```diff
@@ -463,6 +464,22 @@ impl<T: ?Sized> *const T {
     #[stable(feature = "pointer_byte_offsets", since = "1.75.0")]
     #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "1.75.0")]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // If the size of the pointee is zero, then `count` must also be zero
+        (mem::size_of_val_raw(self) == 0 && count == 0) ||
```

Comment:
Updated the contracts to address this.

Reply:
Sorry, I wasn't clear: What I mean is that if `mem::size_of_val_raw(self) == 0`, then *all* counts are valid, because "every pointer is valid". I think something like this would work for `byte_offset`:
```rust
// Per https://doc.rust-lang.org/1.82.0/core/ptr/index.html#safety:
//
//   For operations of size zero, every pointer is valid, including the null pointer.
mem::size_of_val_raw(self) == 0 || 
    // If the referent is non-zero-sized...
    (
        // Per https://doc.rust-lang.org/1.82.0/std/primitive.pointer.html#method.offset:
        //
        //   The computed offset, `count * size_of::<T>()` bytes, must not overflow isize.
        (self.addr() as isize).checked_add(count).is_some()
        // Per https://doc.rust-lang.org/1.82.0/std/primitive.pointer.html#method.offset:
        //
        //   `self` must be derived from a pointer to some allocated object, and the 
        //   entire memory range between self and the result must be in bounds of
        //   that allocated object
        && kani::mem::same_allocation(self, self.wrapping_byte_offset(count))
    )
)
```
I.e., we can exchange the `count == 0` in your latest code for `mem::size_of_val_raw(self) == 0`.

That said, I just tried testing this in miri:
```rust
#[feature(strict_provenance)]

fn main() {
    struct Zst;
    let dangling = std::ptr::dangling::<Zst>();
    // SAFETY: `dangling`'s referent is zero-sized.
    let _ = unsafe { dangling.byte_offset(42) };
}
```
...and got the error:
```
error: Undefined Behavior: out-of-bounds pointer arithmetic: expected a pointer to 42 bytes of memory, but got 0x1[noalloc] which is a dangling pointer (it has no provenance)
 --> src/main.rs:7:22
  |
7 |     let _ = unsafe { dangling.byte_offset(42) };
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^ out-of-bounds pointer arithmetic: expected a pointer to 42 bytes of memory, but got 0x1[noalloc] which is a dangling pointer (it has no provenance)
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
  = note: BACKTRACE:
  = note: inside `main` at src/main.rs:7:22: 7:46

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
```
...so it looks like we've found an inconsistency in Rust's documentation and implementation! Either the implementation of `byte_offset` should look more like this ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a27638a5bfeba004366268eb762a5450)):
```rust
pub unsafe fn byte_offset<T: ?Sized>(ptr: *const T, count: isize) -> *const T {
    let (addr, meta) = ptr.to_raw_parts();
    let addr = addr.map_addr(|addr| addr.wrapping_add_signed(count));
    core::ptr::from_raw_parts(addr as *const (), meta)
}
```
...or the documentation on `byte_offset` should explicitly say that the usual guidance about ZSTs does not apply. 

@celinval, does this sound right to you?

Reply:
I clarified with t-opsem that operation size here refers to offset, not the pointee size.

Reply:
Ok Thank you!